Use ApplicationRecordLog in ApplicationSampler

cmcantalupo opened the pull request:
- Modify test constructor to take filter name and avoid
      environment() call in implementation.
    - Update ApplicationSampler unit tests to use ApplicationRecordLog mock
    - Removed checks for logic about hints related to regions.
    - Added a ApplicationSampler test with two ranks.
    - Add ApplicationSampler test for short region functionality

dianarg committed:
Add ApplicationStatus class

- Provides an interface from the application to the controller
  for process id, hint, and progress information.

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Use cache line constant in ApplicationStatus

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Finish up docs for ApplicationStatus; make getters const

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

fkeceli committed:
Added GEOPM region markup to Pennant.

Signed-off-by: Fuat Keceli <fuat.keceli@intel.com>

fkeceli committed:
Fixed app exec selection based on data set.

Signed-off-by: Fuat Keceli <fuat.keceli@intel.com>

fkeceli committed:
Fixed the Makefile to include the new pennant patch file.

Signed-off-by: Fuat Keceli <fuat.keceli@intel.com>

cmcantalupo committed:
Use ApplicationRecordLog in ApplicationSampler

- Modify test constructor to take filter name and avoid
  environment() call in implementation.
- Update ApplicationSampler unit tests to use ApplicationRecordLog mock
- Removed checks for logic about hints related to regions.
- Added a ApplicationSampler test with two ranks.
- Add ApplicationSampler test for short region functionality
- ApplicationSampler::connect() call from Controller.

Co-developed-by: Diana Guttman <diana.r.guttman@intel.com>
Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

fkeceli committed:
Fixed a region markup in pennant.

Signed-off-by: Fuat Keceli <fuat.keceli@intel.com>

fkeceli committed:
Fixed a region markup in pennant.

Signed-off-by: Fuat Keceli <fuat.keceli@intel.com>

dianarg committed:
Remove get/set_hash from ApplicationStatus

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

cmcantalupo commented on src/ApplicationIO.cpp:
```diff
@@ -91,6 +94,7 @@ namespace geopm
 
             if (m_application_sampler.get_io_sample() == nullptr) {
                 auto rank_per_node = m_application_sampler.get_sampler()->rank_per_node();
+
```
delete change

cmcantalupo commented on src/ApplicationIO.cpp:
```diff
@@ -99,6 +103,7 @@ namespace geopm
 
                 platform_io().get_profileio()->connect();
             }
+            m_application_sampler.connect(m_shm_key);
```
Maybe this call should just go in the controller next to where application_io->connect() is called.

cmcantalupo commented on src/ApplicationIO.hpp:
```diff
@@ -168,6 +169,8 @@ namespace geopm
             std::vector<bool> m_is_epoch_changed;
             bool m_is_connected;
             ApplicationSampler &m_application_sampler;
+            std::string m_shm_key;
```
left over cruft

fkeceli committed:
Removed the second set of markup (high-level) from pennant code.

Signed-off-by: Fuat Keceli <fuat.keceli@intel.com>

dianarg commented on test/ApplicationSamplerTest.cpp:
```diff
@@ -56,179 +65,174 @@ class ApplicationSamplerTest : public ::testing::Test
-        {0,       epoch_id,     {{14,0}},     0.0},
-        {0,       region_id,    {{15,0}},     1.0},
+    uint64_t region_hash = 0xabcdULL;
+    std::vector<record_s> message_buffer {
```
this test could have multiple ranks too

cmcantalupo committed:
Update ApplicationSamplerTest.with_epoch to have two processes

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dianarg reviewed (commented):
LGTM

dianarg committed:
Address review comments

- Fixed order of writes for progress
- Removed early return
- Added more doxygen

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Add region hash methods to ApplicationStatus

- This reverts commit bb473250da06b010fe1d026fa4f67c3c1226df10.

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

bgeltz commented on src/ApplicationIO.cpp:
```diff
@@ -54,16 +54,14 @@ namespace geopm
     constexpr size_t ApplicationIOImp::M_SHMEM_REGION_SIZE;
 
-    ApplicationIOImp::ApplicationIOImp(const std::string &shm_key)
-        : ApplicationIOImp(shm_key,
```
Seems like the shm_key handling here was never used previously?  I'm trying to recall why we plumbed this all in.

dianarg replied:
Maybe we had intended to pass it through and forgot to fix this?  The other objects that use shared memory independently look up the environment variable for the shm_key so it should actually be the same in practice.

bgeltz commented on src/ApplicationSampler.cpp:
```diff
@@ -130,133 +132,57 @@ namespace geopm
+        // Clear the buffers that we will be building.
+        m_record_buffer.clear();
+        m_short_region_buffer.clear();
+        // Itereate over the record log for each process
```
*Iterate

bgeltz reviewed (changes requested):
Nits.

dianarg committed:
Remove ApplicationStatus methods returning vectors

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Rename get_work_progress() to get_progress_cpu()

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

dianarg committed:
Fix signed-unsigned comparison in gtest

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>
